aiida/tools/importexport/dbexport/__init__.py

```diff
@@ -131,23 +134,35 @@ def total_time(self) -> float:
         return (self.time_write_stop or self.time_collect_stop) - self.time_collect_start
 
 
+# should make more generic
+ProgressContext = Callable[..., ContextManager[tqdm]]
+
+
 class ArchiveWriterAbstract(ABC):
     """An abstract interface for AiiDA archive writers."""
 
-    def __init__(self, filename: str, **kwargs: Any):
+    def __init__(self, filename: str, progress_context: Optional[ProgressContext] = None, **kwargs: Any):
         """An archive writer
 
         :param filename: the filename (possibly including the absolute path)
             of the file on which to export.
+        :param progress_context: A context manager for creating and updating a progress bar
```

Comment:
maybe add an example of what one could provide here (`tqdm.tqdm`) and what is expected of this progress bar (which methods does it need to have, ...)

Reply:
yeh this is a bit of a pain. Ideally this "description" would be provided entirely via the typing (i.e. `PROGRESS_CONTEXT`). In theory this can be done by replacing `tdqm` with a `Protocol` class, but this is only available natively from python 3.8 or by the backport https://pypi.org/project/typing-extensions/. So I'm not sure what the feeling is about adding yet another dependency (albeit that it will naturally phase itself out)

For full completness you would also need https://pypi.org/project/mypy-extensions/, to use https://mypy.readthedocs.io/en/stable/additional_features.html#extended-callable-types.

Something like

```python
class ProgressBar(Protocol):
   def set_description_str(self, text: str, refresh: bool = True):
      pass
   def update(self, n: int = 1):
      pass

ProgressContext = Callable[[DefaultArg(str, 'desc'), DefaultArg(int, 'total')], ContextManager[ProgressBar]]

```

Reply:
From my point of view, a textual description will suffice fully here.
I think adding dependencies just for this one instance would be overkill; we can decide to do this at a later point.

Reply:
changed